### PR TITLE
Fix: locks calculation with shards

### LINKED ISSUE #2135: Governance. Governance lock. Calculation is wrong, when already have locked balance.
## Description

When user already has 10 locked tokens in governance and try to delegate 1 more token the app shouldn't show 11 locked tokens, it will be still 10.
Screenshot 2024-08-19 at 14 01 10
Screenshot 2024-08-19 at 14 01 05

## Steps to Reproduce

1. Vote for referenda with 10 WND
2. Try to add to another track delegation with 1 WND

## Expected Behavior
Shows 10 lock tokens with increased lock time

## Actual Behavior
Shows 10 -> 11

## Environment
https://github.com/novasamatech/nova-spektr/commit/a1e90ef296095ad76870c380166686644a2b55d7

### LINKED ISSUE #2184: Governance. Wrong available amount in Delegation
## Description

When user delegate with already locked tokens it shows wrong Available balance (does not include locked balance)
Screenshot 2024-08-22 at 12 32 41

## Steps to Reproduce

1. Add locked balance via vote or delegate
2. Go to Add new delegation flow

## Expected Behavior
On confirmation page shows locked balance + transferable in available part

## Actual Behavior
Only transferable balance part is shown

## Environment
https://github.com/novasamatech/nova-spektr/commit/ee67896d8f551b7d6792cb9880528ac7c66e1bb4

### PR DESCRIPTION
**Pull request by sokolova-an**
- fixed available balance calculation for delegation and vote 
- fixed governance lock balance for accounts with shards

### CODE REVIEW

File: src/renderer/features/governance/utils/getLocksForAccount.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import { type BN, BN_ZERO } from '@polkadot/util';
+
+import { toAddress } from '@/shared/lib/utils';
+import { type Address, type TrackId } from '@shared/core';
+
+export const getLocksForAccount = (
```
Comment (johnthecat) [inline src/renderer/features/governance/utils/getLocksForAccount.ts]: I think it's better to calculate address outside and rename method `getLocksForAddress`

File: src/renderer/widgets/DelegateModal/model/form-model.ts [hunk 2]
```diff
@@ -199,6 +206,10 @@ const $delegateForm = createForm<FormParams>({
       init: '',
       rules: [BondNominateRules.description.maxLength],
     },
+    locks: {
+      init: {} satisfies Record<string, BN>,
```
Comment (johnthecat) [inline src/renderer/widgets/DelegateModal/model/form-model.ts]: It should infer from `FormParams` automatically

File: src/renderer/widgets/DelegateModal/ui/DelegateForm.tsx [hunk 3]
```diff
@@ -260,7 +259,7 @@ const FeeSection = () => {
           </DetailRow>
 
           <DetailRow label={t('governance.locks.governanceLock')} wrapperClassName="items-start">
-            <LockValueDiff asset={network.asset} from={totalLock} to={amountValue} />
+            <LockValueDiff asset={network.asset} from={accounts[0].lock} to={amountValue} />
```
Comment (johnthecat) [inline src/renderer/widgets/DelegateModal/ui/DelegateForm.tsx:262]: `accounts[0]` can return undefined, check it outside JSX

File: src/renderer/widgets/VoteModal/aggregates/voteForm.ts [hunk 4]
```diff
@@ -99,23 +101,34 @@ const transactionForm = createTransactionForm<Form>({
 
 const { form, resetForm, transaction, accounts } = transactionForm;
 
+sample({
+  clock: form.fields.account.onChange,
+  source: {
+    trackLocks: locksModel.$trackLocks,
+    chain: networkSelectorModel.$governanceChain,
+  },
+  filter: ({ chain }, account) => !isNil(account) && !isNil(chain),
```
Comment (johnthecat) [inline src/renderer/widgets/VoteModal/aggregates/voteForm.ts]: `nonNullable(account) && nonNullable(chain)`

File: src/renderer/features/governance/utils/getLocksForAccount.ts (hunk 1 above)
Comment (johnthecat) [reply src/renderer/features/governance/utils/getLocksForAccount.ts]: Also you can move this helpers into lockService

File: src/renderer/features/governance/utils/getLocksForAddress.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { type BN, BN_ZERO } from '@polkadot/util';
+
+import { type Address, type TrackId } from '@shared/core';
+
+export const getLocksForAddress = (address: string, trackLocks: Record<Address, Record<TrackId, BN>>): BN => {
+  return Object.values(trackLocks[address]).reduce((max, x) => (max.gte(x) ? max : x), BN_ZERO);
```
Comment (Asmadek) [inline src/renderer/features/governance/utils/getLocksForAddress.ts]:
```suggestion
  return Object.values(trackLocks[address]).reduce((max, x) => BN.max(max, x), BN_ZERO);
```

**Review by Asmadek — APPROVED** [review]

File: src/renderer/widgets/DelegateModal/model/form-model.ts [hunk 6]
```diff
@@ -59,26 +62,31 @@ const $accounts = combine(
...
 
     const { chain, asset } = network;
 
     return shards.map((shard) => {
       const balance = balanceUtils.getBalance(balances, shard.accountId, chain.chainId, asset.assetId.toString());
+      const address = toAddress(shard.accountId, { prefix: network!.chain.addressPrefix });
+      const lock = getLocksForAddress(address, trackLocks);
 
-      return { account: shard, balance: transferableAmount(balance) };
+      return { account: shard, balance: transferableAmount(balance), lock };
```
Comment (johnthecat) [inline src/renderer/widgets/DelegateModal/model/form-model.ts]: Can we switch balance type to `BN` (using `transferableAmountBN`)?

**Review by johnthecat — CHANGES_REQUESTED** [review]

**Review by johnthecat — APPROVED** [review]

**Pull request merged**